llmfoundry/callbacks/kill_loss_spike_callback.py

```diff
@@ -0,0 +1,198 @@
+        high_loss_count = sum(
+            1 for loss in self.loss_window if loss > self.loss_cap
+        )
+        is_high_loss = (high_loss_count >= self.window_size / 2)
+
+        if is_high_loss:
+            log.info(
+                f'High losses detected: {high_loss_count}/{self.window_size} losses above {self.loss_cap}.',
+            )
+
+        return is_high_loss
+
+    def _log_metadata(self, logger: Logger, key: str, message: str) -> None:
+        for destination in logger.destinations:
+            if isinstance(destination, MosaicMLLogger):
+                destination.log_metadata({
+                    key: message,
+                    'loss_window': list(self.loss_window),
+                })
+
+    def _handle_loss_spike(
+        self,
+        logger: Logger,
+        running_loss_avg: float,
+    ) -> None:
+        message = f'Training loss spike detected for {self.outlier_counter} consecutive steps. Consider stopping this run and resubmitting with a lower learning rate.'
+        self._log_metadata(logger, 'loss_spike', message)
+        if not self.log_only:
+            raise LossSpikeError(
+                outlier_multiplier=self.outlier_multiplier,
+                running_loss_avg=round(running_loss_avg),
+                outlier_counter=self.outlier_counter,
+            )
+
+    def _handle_high_losses(self, logger: Logger) -> None:
+        message = f'Persistently high (>{self.loss_cap}) training losses detected. Consider stopping this run and resubmitting with a lower learning rate.'
+        self._log_metadata(logger, 'high_loss', message)
+        if not self.log_only:
+            raise HighLossError(
+                loss_cap=self.loss_cap,
+                window_size=self.window_size,
+            )
+
+    def _set_window_size(self, state: State) -> None:
+        total_training_steps = 0
+        current_step = int(state.timestamp.batch)
+        current_token = int(state.timestamp.token)
+
+        if state.max_duration is not None:
+            if state.max_duration.unit == TimeUnit.EPOCH and state.dataloader_len is not None:
+                total_training_steps = state.dataloader_len * state.max_duration.value
+            elif state.max_duration.unit == TimeUnit.BATCH:
+                total_training_steps = state.max_duration.value
+            elif state.max_duration.unit == TimeUnit.TOKEN:
+                # This is an approximation of the total batches from the total tokens, assuming the ratio of tokens:batch is constant.
+                total_training_steps = current_step * (
+                    state.max_duration.value / current_token
+                )
+            self.window_size = max(
+                self.window_size,
+                round(float(total_training_steps * _WINDOW_FRACTION)),
+            )
+        self.loss_window = deque(self.loss_window, maxlen=self.window_size)
```

Comment:
rather than this checking of whether window size has been initialized, I believe you can just use the `fit_start` event to set loss window appropriately before the first batch. you can initialized loss window to none in the constructor.

Reply:
I think we need some history of batches in order to set the correct loss window if max_duration is given in tokens? And then it was just a decision to keep all the window operations in the same place (I can break the epoch/batch ones off into fit_start if you think that's better)

Reply:
oh yeah, you're right. all good